### PR TITLE
torch/_native: OpInfo coverage for the chunked linear_cross_entropy overrides

### PR DESCRIPTION
**Pull request by pearu**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.13.0) (oldest at bottom):
* #195830
* #196134
* #195829
* __->__ #195334
* #195332
* #195331
* #194510
* #194509

The overrides should be tested by the existing OpInfo machinery rather than by
hand-written tests, so we add `cutedsl` variants of the two chunked ops whose
samples are the eager generators' samples filtered by the op's own eligibility
predicate. These variants exist only where the CuTeDSL runtime is installed, so
we also add them to the B200 smoke job, which is the only job that has the
runtime.

> `torch/_native/README.md` names an OpInfo entry in `dsl_ops_by_dsl` as the
> preferred way to test an override, so that the existing `test_ops` /
> `test_meta` / gradient / functorch infrastructure applies. This adds one
> variant per chunked op, `cutedsl` and `cutedsl_none`.
>
> The samples are the eager generators' whole space, filtered rather than
> hand-picked, so the variants track those generators instead of drifting from
> them. `_cutedsl_linear_cross_entropy_eligible` is the filter, and it asks
> `F.linear_cross_entropy`'s own predicate -- extracted in the parent commit for
> this caller -- whether a chunked op is reached, so the selection cannot drift
> from the dispatch it is predicting. Both variants share the filter, because
> `reduction` selects *which* of the two ops runs rather than whether either does,
> and each generator already fixes that. Nothing an op receives is filtered out.
>
> Sharing the predicate removes the risk of the filter drifting, but not of the
> filter handing it the wrong pieces of a `SampleInput`, nor of the predicate
> disagreeing with what the call actually does. So
> `test_sample_filter_matches_the_gate` checks both across each generator's full
> space, in both grad modes, and that the filter admits something. The signal it
> uses is that the shared accumulator runs if and only if a chunked op does -- the
> reference path uses `linear` + `cross_entropy` instead -- so its call count
> answers exactly the question the predicate answers, with no dispatcher
> introspection.
>
> Two details there are load-bearing. The mirror is evaluated in the same grad
> mode as the call it predicts, because one of its clauses reads
> `torch.is_grad_enabled()` and evaluating it under a different mode would make
> the two disagree by construction. And both grad modes are covered, since
> eligibility genuinely differs between them -- a grad-requiring probability
> target goes to the reference path -- and that clause is unreachable otherwise.
>
> Exceptions from the call are not tolerated. On supported hardware no sample of
> either generator raises, in either grad mode, so catching them could only hide
> a regression: an override that started throwing would be counted and skipped
> rather than reported.
>
> Everything else in the two entries is copied from the eager `chunked` and
> `chunked_none` variants, because the behaviour is the op's rather than the
> override's: dtypes, `gradcheck_nondet_tol`, the COW materialization allow-lists,
> the inherited autograd limits, and the JIT / Inductor / decomposition skips.
> What differs is only what the override imposes -- CUDA only, CuTeDSL required,
> and the filtered samples. `cutedsl_none` additionally takes
> `check_batched_grad=False`, `check_batched_gradgrad=False` and the
> vmap-over-grad expected failure, mirroring `chunked_none`, whose backward
> recomputes from saved inputs.
>
> Three declarations live in the test files rather than the OpInfo, next to the
> eager variants' equivalents: the `test_fn_gradgrad` xfails in
> test_ops_gradients.py, and the `TestCompositeCompliance::test_backward` skip in
> test_ops.py that `chunked_none` already carries (its recompute-in-backward does
> an in-place `mul_` that is not composite compliant).
>
> The three variants' skip lists overlapped by fourteen declarations -- the
> limits any implementation of the chunked op inherits, whatever computes the
> loss: no Inductor lowering, an options dataclass JIT cannot script,
> `test_cow_input`'s unspecified launch failure, and autograd support stopping at
> a first-order VJP. Those move into `_lce_chunked_op_skips`, which the eager
> `chunked` / `chunked_none` entries and both `cutedsl` entries now build on, so
> the inherited set has one definition instead of three.
>
> What stays per variant is what actually differs, and that distinction matters
> more than it looks: an over-applied `expectedFailure` reports an UNEXPECTED
> SUCCESS, so it is not safe to inherit one. `test_jvp`,
> `test_normalize_operator_exhaustive` and `test_vmap_autograd_grad` are
> therefore declared only where the failure was observed -- the first two on the
> eager variants, the third on `chunked_none` alone -- and the MPS accuracy and
> NNC entries stay with the eager variants, which are the ones that run there.
>
> The shared generator also gains two large-logit shapes. Every sample it built
> until now takes its tensors from the shared linear generator, which tops out
> around |z| = 20 with a median of 6 -- and accuracy on this op can depend on |z|
> well past that, because an implementation may hold the logits at low precision
> (eager does for fp16) and such a buffer stores `z` with absolute error
> proportional to `|z|` while the softmax depends on DIFFERENCES of `z`. The two
> shapes raise the magnitude two ways that separate those effects. One puts a
> uniform `linear_bias` over deliberately mild logits -- narrow input and weight
> ranges, so the spread stays a few units while the bias carries |z| into the
> hundreds -- which leaves the softmax mathematically unchanged and isolates
> storage rounding. The other uses a wide weight range with the target at the
> argmax, which is genuinely peaked. Both use class-index targets, like the rest
> of this generator's chunked samples. They are
> emitted in whatever flavour the call asks for -- chunked options for the chunked
> variants, `reduction='none'` for that variant, none for the reference -- so
> every variant sees them, including the eager ones, whose fp16 path has the same
> exposure and nothing else that exercises it.
>
> The variants also need a job that collects them. They are constructed only when
> the CuTeDSL runtime is importable at a version registration accepts -- a variant
> built where `cu.register_op_override` drops every registration would run eager
> under a `cutedsl` label -- and smoke_b200 is the one job that has it --
> and its include list held no test that iterates `op_db`, so until now the
> `cutedsl` and `cutedsl_none` entries ran nowhere in CI, and neither did the
> routing test. Both join that list. `OPINFO_RESTRICT_TO_DSL=cutedsl` narrows
> `op_db` to the DSL entries and `-k` narrows to this op, so the run is the two
> variants and nothing else: 27 tests in 37s on an H100. Keeping it to this op
> means a failure there points at this stack rather than at another DSL override.
>
> Two more suites join it for the same reason. The variants' expectations are not
> confined to `test_ops`: `test_ops_gradients.py` carries two xfails and the
> shared skip tuple names nine `TestOperators` entries, and an expectation that
> never runs rots into an unexpected success unnoticed. The gradient run is
> restricted the same way; functorch's is selected with `-k "linear_cross_entropy
> and cutedsl"` instead, because restricting `op_db` trips `opsToleranceOverride`,
> which asserts that every op it names is present.
>
> #193606, open at the time of writing, replaces that kind of enumeration with
> discovery of `test/python_native/**`. When it lands, the `--include
> python_native/test_linear_cross_entropy_override` line here becomes redundant
> and should be deleted; the `OPINFO_RESTRICT_TO_DSL=cutedsl ... --include
> test_ops` line must STAY, because the OpInfo variants live in `test_ops`, which
> that discovery does not reach, and they exist only where the runtime does. Both
> patches edit `test_python_smoke_b200()` around the same anchor -- this one
> immediately before the `test_linalg -k "mm or addmv"` line, #193606 immediately
> after it -- so whichever lands second wants a manual rebase there rather than a
> clean automatic one.
>
> Test Plan:
>
> Verified on an H100 (sm_90). Each declaration above was added only after
> observing the corresponding failure there -- not predicted -- which is what
> kept `test_vmap_autograd_grad` off the `cutedsl` variant, where it passes.
>
> The B200 job itself can only run in CI, so what was verified locally is that
> its four new invocations select what they claim: 7 tests for the routing file,
> 27 for the DSL-restricted `test_ops` run, 2 passed and 2 xfailed for the
> gradient run, and 2 passed and 20 xfailed for functorch's.
>
> The large-logit samples were run through every battery that consumes this
> generator, on an H100: `test_ops`, `test_ops_gradients` (its fp64 gradcheck
> legs, where a peaked softmax was the plausible risk), `test_meta`, `test_nn`,
> `test_ops` restricted to the cutedsl variants, and functorch's `test_ops` --
> all clean, and no unexpected successes among the 60 expected failures there.
>
> The shared-tuple extraction is behaviour-preserving by construction and was
> checked as such: every variant's resolved skip set was dumped before and after
> the move -- test class, test name, device, dtypes, active_if and decorator kind
> -- and the two dumps are identical, so no variant gained or lost a
> declaration.
>
> ```
> OPINFO_RESTRICT_TO_DSL=cutedsl python -m pytest test/test_ops.py -q -k linear_cross_entropy
> OPINFO_RESTRICT_TO_DSL=cutedsl python -m pytest test/test_meta.py -q -k linear_cross_entropy
> OPINFO_RESTRICT_TO_DSL=cutedsl python -m pytest test/test_decomp.py -q -k linear_cross_entropy
> OPINFO_RESTRICT_TO_DSL=cutedsl python -m pytest test/test_ops_fwd_gradients.py -q -k linear_cross_entropy
> OPINFO_RESTRICT_TO_DSL=cutedsl python -m pytest test/inductor/test_torchinductor_opinfo.py -q -k linear_cross_entropy
> python -m pytest test/test_ops_gradients.py -q -k "linear_cross_entropy and cutedsl"
> python -m pytest test/functorch/test_ops.py -q -k "linear_cross_entropy and cutedsl"
> python -m pytest test/python_native/test_linear_cross_entropy_override.py -q
> python -m pytest test/test_ops.py test/test_meta.py -q -k linear_cross_entropy
> python -m pytest test/python_native/ -q
> python scripts/lintrunner.py
> ```
>
> Restricted to the new variants: 27 passed and 47 skipped in test_ops, 26 and 78
> in test_meta, 4 xfailed and 8 skipped in fwd-gradients, with the decomposition
> (16) and Inductor (12) suites skipped as declared. Unrestricted but filtered to
> them: 2 passed and 2 xfailed in gradients, 2 passed and 20 xfailed in functorch.
> The routing tests pass (7). Across the whole `test/python_native/` directory 287
> pass, and the 8 failures are pre-existing and unrelated: six subtest failures
> plus `test_prod_override_engaged` in test_sum_cutedsl.py, which needs
> `PYTORCH_SUM_INNER_TREE=1` for the sum/prod overrides to register, and
> `test_import_does_not_load_cutlass`, which holds only when its file runs alone.
>
> Unrestricted, against the same checkout without this commit: test_ops 90 -> 117
> passed, test_meta 75 -> 101.
>
> The filter agrees with the dispatch on every sample of all four
> generator/grad-mode combinations, with 92, 52, 52 and 28 of 161, 161, 91 and 91
> samples eligible.

---
_🤖 Drafted by Claude Code (an AI agent) and reviewed & approved by pearu._

### CODE REVIEW

**Comment by slayton58** [comment]
Per codex review - same version gate should be fixed here:

> no new correctness blocker.
>
>  Concern: CI only runs test_ops for the new CuTeDSL variants (.ci/pytorch/test.sh:517). The added gradient expectations and claimed meta/functorch coverage are never continuously collected because those variants exist only in CuTeDSL environments.
>
>  The prior version-gating concern also carries over: standalone OpInfo runs can exercise eager fallback when CuTeDSL is present but unsupported, although the paired routing test catches this in the B200 job.

**Review by slayton58 — APPROVED** [review]
No remaining issues - looks good

**Comment by pearu** [comment]
Both collected now, and the version gate is applied where the variants are built. Detail below, drafted by Claude Code:

> **CI collection.** The B200 job gains two lines beyond `test_ops`: `OPINFO_RESTRICT_TO_DSL=cutedsl ... --include test_ops_gradients -k linear_cross_entropy`, and functorch's `test_ops` selected with `-k "linear_cross_entropy and cutedsl"`. functorch needs the `-k` form rather than the env var: restricting `op_db` trips `opsToleranceOverride`, which asserts that every op it names is present. Measured at this commit on an H100 -- routing file 7 tests, `test_ops` 27, `test_ops_gradients` 2 passed and 2 xfailed, functorch 2 passed and 20 xfailed.
>
> **The version gate.** The OpInfo entries are now constructed only when `cutedsl_utils._version_is_ok()`, so a standalone run cannot exercise eager under a `cutedsl` label. Checked both ways on an H100 by dropping the installed version out of `_CUTEDSL_REQUIRED_VERSIONS`: with the gate open `op_db` holds the `cutedsl` and `cutedsl_none` variants, with it closed it holds neither and the DSL filter reports 3 OpInfos instead of 5.
>
> **What was not narrowed.** The variants keep the op's dtypes rather than the kernel's fp16/bf16. Narrowing them makes `test_dtypes` fail -- it compares the claim against what actually runs -- and empties `test_ops_gradients`, whose gradcheck is float64-only, taking `OK (expected failures=2)` to `NO TESTS RAN`. A comment at `_cutedsl_lce_kwargs` now records that, so a `cutedsl` variant means "the entry whose samples the op's chunked gate filters", not "every sample reached a kernel".
